### PR TITLE
Replaced --forked in internal tests

### LINKED ISSUE #107: Replace --forked in internal tests
[Update: replaced link to comment with full description.]

## Problem

The test suite is run twice on azure: once "normally", and once using `--forked`. But:

1. We no longer need to run a "forked" version of the tests, i.e. we no longer need a 
 version of the test suite that guarantees one test per process (i.e. one cell execution per kernel).

2. `--forked` does not work on windows.

## Proposed solution

* Remove any special handling in the test suite (i.e. within `nbcelltests/tests/`) for the "forked" case.

* Fix `'pytest-xdist ; sys_platform != "win32"'` in setup.py to always install pytest-xdist no matter the platform (I believe it works on windows, just some options e.g. `--forked` don't work/have meaning on windows).

* Replace the `pytest --forked` invocation of the test suite with probably `pytest -n`. That likely means editing `Makefile` and `.azure/pipelines.yml` and `.azure/dev-template.yml`.

Working on the test suite of a testing tool unfortunately likely requires some knowledge of python's test tools (unittest, pytest). Please feel free to ask if you see something that looks weird - it might be just bad code, or it might be necessary for some awkward reason, and it's not always easy to tell.

## Background

`--forked` was originally added to the test suite so that we could run two versions of the test suite:
1. the original way, with multiple tests in the same process (i.e. multiple tests sharing a kernel)
2. with one test per process (i.e. one kernel per test), to identify and fix issues with kernel state (e.g. multiple executions of the same cell in a given kernel).

The full test suite originally only passed in the second case. You can see an example of a test that failed in the first case in test_test.py; the test below originally passed _only_ in the forked case, because only in the forked case were cells not repeatedly executed in the same kernel:

```python
class TestCumulativeRun(_TestCellTests):
    [...]
    def test_state(self):
        """In the expected case, the five cells of the notebook should work
        out as follows:

             cell    test   state
        1:    -       -       -
        2:   x=0      -      x=0
        3:   x+=1     -      x=1
        4:   x+=1     -      x=2
        5:    -      x+=1    x=3

        However, these tests originally detected repeated execution of
        cells in the same kernel. Cell 2 actually only sets x to 0 if
        x is not already defined; if x is already defined, cell 2 does
        not change x's value. If for some reason each test results in
        multiple executions of the cells, the results will look like
        this:

             cell    test   state
        1:    -       -       -
        2:    -       -      x>0   (bad; x was previously defined in kernel)
        3:   x+=1     -      x>1   (bad)
        4:   x+=1     -      x>2   (bad)
        5:    -      x+=1    x>3   (bad)
        """
```

So, the `--forked` version of the test suite was not there to catch a bug - the bug was caught by the regular test suite - but to demonstrate the reason for the regular test suite failing. The bug is long since fixed, so it's fine to remove the `--forked` version and handling for it.

However, it's still useful to run a distributed as well as non-distributed version of the test suite in order to increase the chance of catching certain bugs, hence the suggestion to replace the `--forked` invocation with `-n` (with just some reasonable `n`, e.g. 2 or whatever).

See https://github.com/jpmorganchase/nbcelltests/pull/90#discussion_r423704977 for original discussion.

### PR DESCRIPTION
**Pull request by utsavm9**
Resolves #107:

- `pytest-xdist` will now be installed on Windows
- Replaced `--forked` with `-n 2`
- Tests like `test_basic_runWithReturn_pass` and 4 others do not want to have the file `__cell_coverage_test.py` already created. I grouped all these 5 tests under a class `TestCoverage` so that these tests are run sequentially by a worker spawn by `pytest-xdist`
- Removed special handling of `--forked`. Now, tests like `test_state` behave as if it's always `FORKED=False`.

Naming Choices:
- Would you want to rename `make pytest-forked` to something else?
- Azure test `pytest --forked` renames to `pytest -n 2`
- Is the class name `TestCoverage` apt for grouping these tests? `test_basic_runWithReturn_pass`, `test_basic_runWithReturn_fail`, `test_basic_runWithReport_pass`, `test_basic_runWithHTMLReturn_pass`, `test_basic_runWithHTMLReturn_fail`

It is not apparent looking at the diff of `test_test.py TestCoverage`, but I have only added those five tests to the class and I have not changed those tests.

### CODE REVIEW

**Comment by timkpaine** [comment]
these are failing lint, run `make fix` for auto fix
Screen Shot 2020-09-23 at 01 55 20

File: .azure/dev-template.yml
Comment (ceball) [inline .azure/dev-template.yml:57]: Maybe just `pytest distributed` or `pytest parallelized` or similar. And the same name for the Makefile target.

File: Makefile
Comment (ceball) [inline Makefile:14]: why `dist=loadscope`?

* it's different from the invocation on azure (ideally azure would be using the makefile targets, but that's another story...)
* are we having to rely on some ordering or grouping for the distributed case?

(i.e. I would prefer no guarantees about ordering/grouping - I'd like the distributed case to be least like the regular case)

**Review by ceball — COMMENTED** [review]
This is great, thanks so much! 

Thanks also for making your change so easy to review.

> Would you want to rename make pytest-forked to something else?

Yes, I agree - see comment inline (i.e. same name as used in azure)


We'll be happy to accept this PR, but before merging I want to discuss two aspects. One is covered in the comments (i.e. "distribution algorithm"); the other is:

> Tests like test_basic_runWithReturn_pass and 4 others do not want to have the file `__cell_coverage_test.py` already created. I grouped all these 5 tests under a class `TestCoverage` so that these tests are run sequentially by a worker spawn by pytest-xdist

Please can you explain again why you are grouping these? I can see 3 possible reasons: (a) reduction of duplicated code, (b) grouping genuinely related tests together, i.e. in a "semantically meaningful way", (c) because otherwise the tests don't pass e.g. in the distributed case.

I can't quite tell which of those reasons (or maybe another?) applies here from your comment. 


> Is the class name TestCoverage apt for grouping these tests

Ideally I would like to group only if it makes real sense to do so, not if it's fixing some problem we could solve another way. So the answer to this one depends on the outcome of the discussion above.

File: Makefile
Comment (utsavm9) [reply Makefile:14]: - I do not think it is different from the invocation on Azure. I think it just calls`make testpy-distributed` now which will come to Makefile and run the command with `--dist=loadscope`. Does it not?
- Those 5 tests need to be sequential.

**Comment by utsavm9** [comment]
> Please can you explain again why you are grouping these? 

You guessed it right, without grouping the tests will not always pass. Those tests first check if `__cell_coverage_test.py` has not already been created and fail if it already exists. The file would be present if another of those 5 tests is running in another worker. https://github.com/jpmorganchase/nbcelltests/blob/aef8e3274edd8ec6ed84e0866b555c6e72c4a50b/nbcelltests/tests/test_test.py#L690

---

So, those 5 tests cannot be parallelized. One way to make `pytest-xdist` run those sequentially is to group them in a class and pass `--dist=loadscope`. From the [docs](https://pypi.org/project/pytest-xdist/):

> `--dist=loadscope`: Tests are grouped by module for test functions and by class for test methods. Groups are distributed to available workers as whole units.

**Comment by ceball** [comment]
Quick comment from my phone...

I think those tests (or the underlying functions they test) should be fixed. E.g. they should use a unique temporary file.

We should either do that here in this PR, or open an issue to do it in the future. Up to you! (Off the top of my head I don't know if it's a big change or not.)

**Review by ceball — COMMENTED** [review]
> I do not think it is different from the invocation on Azure. I think it just callsmake testpy-distributed now which will come to Makefile and run the command with --dist=loadscope. Does it not?

Ok, great. (Likely I misread the diff :) )

**Comment by utsavm9** [comment]
I am not sure at all why setup of [TestSkips is failing on Linux Python37](https://dev.azure.com/tpaine154/jupyter/_build/results?buildId=1787&view=logs&j=2a28960a-ed8d-5519-388d-3ffe4ab60c96&t=2bc57bbe-2684-594b-9fba-5c9a3ce4243c&l=303). I am trying to duplicate that on my local machine.

Let's open another issue to make those tests independent. There are two files involved: `__cell_coverage_test.py` and `_cell_coverage_test.py`. I can get back to that after a while.

**Comment by ceball** [comment]
> I am not sure at all why TestSkips is failing to setup once on Linux Python37. I am trying to duplicate that on my local machine.

Looked like a timeout to me, probably just from azure. So I re-ran, and it passed. 

(If it was down to some bad interaction between the tests that only shows up sometimes in the distributed case, we'll address it if we see it again in the future...)

**Comment by utsavm9** [comment]
One last thing to ask: Would you like to be able to do `make testpy-distributed N=8` along with `make testpy-distributed`? I personally have been using `N=4` on my university's Linux server to test my changes as that was the fastest on that machine. See this: https://github.com/utsavm9/nbcelltests/commit/c934476b3cc8e89bc210c33c6bb7d7a0bab86bb5 

Apart from that, I am all done with this PR!

**Comment by ceball** [comment]
> Let's open another issue to make those tests independent. There are two files involved: __cell_coverage_test.py and _cell_coverage_test.py, so it will take a while and I can get back to it after some time.

Are you able to make the issue (extracting and summarizing our discussion here)? (No obligation; if not, I will do it.)

**Comment by ceball** [comment]
> One last thing to ask: Would you like to be able to do make testpy-distributed N=8 

Please could you make that as a separate PR, after this one? I am not at all against adding an option for varying n when running the tests (I think it's useful), but I don't own the Makefile so would want a teammate to ok it.

File: Makefile
Comment (ceball) [reply Makefile:14]: Please add a comment here explaining why loadscope (and ideally would link to a github issue).

**Comment by utsavm9** [comment]
OK, I will make another PR to master after this PR is merged. Thanks!

File: nbcelltests/tests/test_test.py
Comment (ceball) [inline nbcelltests/tests/test_test.py:725]: Please add same comment here as in Makefile (i.e. explain why these tests are grouped into a class, linking to the same github issue if you create the issue).

**Review by ceball — CHANGES_REQUESTED** [review]
Just requested some comments (and ideally a new issue), otherwise this PR is fine with me. Thanks!

File: nbcelltests/tests/test_test.py
Comment (ceball) [reply nbcelltests/tests/test_test.py:725]: (Sorry, the above comment should have been on the line below, i.e. it's referring to the new `TestCoverage` class.)

File: Makefile
Comment (ceball) [inline Makefile:15]:
```suggestion
# required until https://github.com/jpmorganchase/nbcelltests/issues/207 is addressed
```

**Review by ceball — APPROVED** [review]
Thanks for all the follow-up work!

**Comment by ceball** [comment]
I'll merge once the tests pass (can't merge until then).

**Pull request merged**
